src/Conversion/ONNXToKrnl/Tensor/DepthToSpace.cpp

```diff
@@ -0,0 +1,185 @@
+    (void)shapeComputed;
+    assert(succeeded(shapeComputed) && "Could not compute output shape");
+
+    Value input = spaceToDepthOp.input();
+    int64_t bs = spaceToDepthOp.blocksize();
+    Location loc = spaceToDepthOp.getLoc();
+    StringRef mode = spaceToDepthOp.mode();
+
+    // Compute the new dimensions.
+    MemRefBoundsIndexCapture bounds(input);
+    assert(bounds.getRank() == 4 && "Input tensor should have rank equal to 4");
+
+    DimIndexExpr B(bounds.getDim(0));
+    DimIndexExpr C(bounds.getDim(1));
+    DimIndexExpr H(bounds.getDim(2));
+    DimIndexExpr W(bounds.getDim(3));
+    DimIndexExpr newC = create<SignedFloorDivIOp>(C, (bs * bs), rewriter, loc);
+    DimIndexExpr newH = create<MulIOp>(H, bs, rewriter, loc);
+    DimIndexExpr newW = create<MulIOp>(W, bs, rewriter, loc);
+
+    // Reshape input tensor to shape [B, bs, bs, C/(bs*bs), H, W].
+    LiteralIndexExpr bsLit(bs);
+    DimsExpr outputDims1;
+    ArrayRef<int64_t> perm;
+    if (mode == "DCR") {
+      outputDims1 = {B, bsLit, bsLit, newC, H, W};
+      perm = {0, 3, 4, 1, 5, 2};
+    } else {
+      assert(mode == "CRD" && "Unexpected mode");
+      outputDims1 = {B, newC, bsLit, bsLit, H, W};
+      perm = {0, 1, 4, 2, 5, 3};
+    }
+    Value reshapeRes1 = reshape(input, outputDims1, rewriter, loc);
+    LLVM_DEBUG(dbgs() << "reshapeRes1: " << reshapeRes1 << "\n");
+
+    // Transpose the reshape result into shape [B, C/(bs*bs), H, bs, W, bs].
+    DimsExpr outputDims2({B, newC, H, bsLit, W, bsLit});
+    Value transposeRes = transpose(
+        reshapeRes1, outputDims2, perm, rewriter, spaceToDepthOp, shapeHelper);
+    LLVM_DEBUG(dbgs() << "transposeRes: " << transposeRes << "\n");
+
+    // Reshape the transpose result into shape [B, C/(bs*bs), H*bs, W*bs].
+    DimsExpr outputDims3({B, newC, newH, newW});
+    Value reshapeRes2 = reshape(transposeRes, outputDims3, rewriter, loc);
+    LLVM_DEBUG(dbgs() << "reshapeRes2: " << reshapeRes2 << "\n");
+
+    rewriter.replaceOp(op, reshapeRes2);
+
+    return success();
+  }
+
+private:
+  template <typename Op>
+  DimIndexExpr create(const DimIndexExpr &indexExpr,
+      const LiteralIndexExpr &constant, ConversionPatternRewriter &rewriter,
+      const Location &loc) const {
+    static_assert(Op::template hasTrait<OpTrait::NOperands<2>::Impl>(),
+        "expected binary operation");
+
+    if (!indexExpr.isLiteral())
+      return DimIndexExpr(
+          rewriter.create<Op>(loc, indexExpr.getValue(), constant.getValue()));
+
+    if (std::is_same<Op, SignedFloorDivIOp>::value)
+      return LiteralIndexExpr(indexExpr.getLiteral() / constant.getLiteral());
+    else if (std::is_same<Op, MulIOp>::value)
+      return LiteralIndexExpr(indexExpr.getLiteral() * constant.getLiteral());
+
+    llvm_unreachable("Unexpected operator type");
+  }
+
+  // Reshape the 'input' tensor to the shape prodided by 'outputDims'.
+  Value reshape(const Value &input, const DimsExpr &outputDims,
```

Comment:
Could we call ONNXReshape instead of manually doing the lowering? We want to avoid the same lowering at multiple places.